### PR TITLE
Fractions functionality issue #395

### PR DESCRIPTION
**Pull request by SCMusson**
Targetting issue #395.
Dunder methods for all Fractions functions.
The Fraction functions are still there, I don't know if you want to temporarily keep them for backwards compatibility or to be replaced entirely with dunder methods.

Also created `Fraction.norm` and `Fraction.ceil` class methods which could replace `ceil_fraction` and `norm_fraction`.

### CODE REVIEW

**Comment by nielstron** [comment]
I would remove the old methods.

One plan for this was also that it is merged after #397, such that add mul etc have support for Union[int, Fraction].

**Comment by SCMusson** [comment]
That's fine.  #397 is pretty much ready to go. I'll come back to this when its merged.

**Comment by SCMusson** [comment]
I had to modify a few different things so that Self in Unions was accepted. 

Currently running something like:

```
from opshin.tests.utils import eval_uplc
from opshin.std.fractions import Fraction
a = Fraction(5,7)
b = Fraction(11,13)

source_code = """
from opshin.std.fractions import *
from typing import Dict, List, Union

def validator(a: Fraction, b: Union[Fraction, int]) -> Fraction:
    return a + b
"""
ret = eval_uplc(source_code, a,b)
```
Works but running `eval_uplc(source_code, a, 5)`

Returns the following error:
```
Traceback (most recent call last):
  File "/home/wppj21/Workshop/opshin/test_class.py", line 121, in <module>
    ret = eval_uplc(source_code, a, 5)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/wppj21/Workshop/opshin/opshin/tests/utils.py", line 35, in eval_uplc
    raise ret
  File "/home/wppj21/miniconda3/envs/opshin/lib/python3.11/site-packages/uplc/machine.py", line 111, in eval
    stack.append(self.return_compute(step.context, step.value))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/wppj21/miniconda3/envs/opshin/lib/python3.11/site-packages/uplc/machine.py", line 171, in return_compute
    return self.apply_evaluate(context.ctx, context.val, value)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/wppj21/miniconda3/envs/opshin/lib/python3.11/site-packages/uplc/machine.py", line 210, in apply_evaluate
    res = eval_fun(*arguments)
          ^^^^^^^^^^^^^^^^^^^^
  File "/home/wppj21/miniconda3/envs/opshin/lib/python3.11/site-packages/uplc/ast.py", line 745, in wrapped_fun
    assert isinstance(
           ^^^^^^^^^^^
AssertionError: Argument 0 has invalid type, expected type <class 'uplc.ast.BuiltinInteger'> got <class 'uplc.ast.PlutusInteger'> (PlutusInteger(value=5))
```

Is this an bug in uplc or have I made an error somewhere?

**Comment by nielstron** [comment]
There is a bug in OpShin, likely something we overlooked in #397.

When something is a Union[x,y] it is PlutusData. However for int and other primitive types I usually assume they are unwrapped into builtin data. Here there must be some missing unwrapping of int after casting it via isinstance. Note that addition can not be done with PlutusData and requires unwrapping via UInt before.

**Comment by nielstron** [comment]
Try

```
def validator(x: Union[int, A]):
  if isinstance(x, int):
     print(x+2)

**Comment by nielstron** [comment]
There is likely also another edge case in #397

```
def validator(x: int):
   y : Union[int, A] = x
   if isinstance(y, int):
     print("foo")
```

will likely crash

**Comment by SCMusson** [comment]
Is CI not working at the moment?

**Comment by nielstron** [comment]
Great, thanks for the update. I think Travis is currently broken, I wanted to migrate to Github actions anyways and am preparing #406 for this. Will merge it into this branch to check that the tests are now still passing.

**Comment by nielstron** [comment]
[It looks like Self is a 3.11 only feature of typing.](https://docs.python.org/3.11/library/typing.html#typing.Self) In previous python versions this is instead solved using literal strings like "Fraction".

**Comment by SCMusson** [comment]
Ah, I'll rewrite them all as literal strings then.

File: opshin/tests/test_Unions.py [hunk 1]
```diff
@@ -305,3 +305,22 @@ def validator(x: Union[int, bytes, bool]) -> int:
...
+from dataclasses import dataclass
+from typing import Dict, List, Union
+from pycardano import Datum as Anything, PlutusData
+
+def validator(x: Union[int,bytes]) -> int:
+    k: int = 0
+    if isinstance(x, int):
+        k = x+5
+    elif isinstance(x, bytes):
+        k = 7
```
Comment (nielstron) [inline opshin/tests/test_Unions.py:321]:
```suggestion
        k = len(x)
```

File: opshin/tests/test_Unions.py [hunk 2]
```diff
@@ -305,3 +305,22 @@ def validator(x: Union[int, bytes, bool]) -> int:
...
+    return k
+"""
+        res = eval_uplc_value(source_code, x)
+        real = x + 5 if isinstance(x, int) else 7
```
Comment (nielstron) [inline opshin/tests/test_Unions.py:325]:
```suggestion
        real = x + 5 if isinstance(x, int) else len(x)
```

**Comment by nielstron** [comment]
The fraction code looks great now. Unfortunately I still found (and added) 2 failing cases for #397.

**Comment by SCMusson** [comment]
I'll get to work on those

**Comment by SCMusson** [comment]
Thank you for those test cases, the first one I really should have thought of myself but I wouldn't have thought of the test_Union_builtin_cast_direct. I had to modify them so that the tests were correct so It might be worth you checking to make sure I didn't break the test.

**Comment by nielstron** [comment]
Your modifications of if-expressions made me think... caught another bug :)

**Comment by nielstron** [comment]
Sorry about the never ending new crashing examples... the last ones added are out of scope though (since not even properly implemented for unions of plutus data), I created #408 dedicated for them. also its not a problem if an expression does not compile.

File: tests/test_Unions.py [hunk 3]
```diff
@@ -305,3 +305,182 @@ def validator(x: Union[int, bytes, bool]) -> int:
...
+        k = foo([B(b"0"*x)])
+    return k[0]
+"""
+        res = eval_uplc_value(source_code, x)
+        real = x + 1 if x > 5 else len(b"0" * x)
+        self.assertEqual(res, real)
+
+    @unittest.skip("Throw compilation error, hence not critical")
+    @hypothesis.given(st.sampled_from(range(14)))
+    def test_Union_builtin_cast_ifexpr(self, x):
```
Comment (nielstron) [inline tests/test_Unions.py:465]:
```suggestion
    def test_Union_builtin_cast_List(self, x):
```

**Comment by nielstron** [comment]
Any update on this? :)

**Comment by SCMusson** [comment]
Apologies, I almost forgot I was supposed to be working on this. I've got the IfExp working now

**Comment by nielstron** [comment]
Great, thank you! This will be part of the next version 0.24.0. Please submit your wallet address for the bug bounty :)

**Comment by SCMusson** [comment]
If this also signals the completion of #397 the same address can be used for the payout of that bounty too.

**Comment by nielstron** [comment]
Paid out both #367 and #395 
https://adastat.net/transactions/c62300dc553ae877376c7bb281997faf453cd5776c16db4e75b14584efbc76ce

**Pull request merged**